ruby/ql/lib/codeql/ruby/frameworks/core/Array.qll

```diff
@@ -1815,6 +1825,30 @@ module Array {
       preservesValue = true
     }
   }
+
+  private import codeql.ruby.frameworks.core.Kernel
+
+  /**
+   * Holds if there an array element `pred` might taint the array defined by `succ`.
+   * This is used for queries where we consider an entire array to be tainted if any of its elements are tainted.
+   */
+  predicate taintedArrayObjectSteps(DataFlow::Node pred, DataFlow::Node succ) {
```

Comment:
I think the better way to do this is to allow implicit array reads at sinks. We could either do this for the two relevant queries, using `Configuration::allowImplicitRead`, or for all taint tracking configurations, using `defaultImplicitTaintRead`, like e.g. [in Java](https://github.com/github/codeql/blob/2cd1e09a7ea17d796fcb9e31926fc0830d9fd874/java/ql/lib/semmle/code/java/dataflow/internal/TaintTrackingUtil.qll#L178).

Reply:
Isn't this because the kernel method `Array` [does not work](https://ruby-doc.org/3.2.1/Kernel.html#method-i-Array) like you intend in your examples? But I also don't think we currently have flow summaries for that method; I guess it should be something like
```
// already an array
input = "Argument[0].WithElement[0..]" and
output = "ReturnValue"
or
// not already an array
input = "Argument[0].WithoutElement[0..]" and
output = "ReturnValue.Element[0]"
```

Reply:
You're completely right. I was looking at how it was used in the CVE, and probably just assumed that it was similar to the JS `Array()` call.  
I've added a summary for `Array()` based on your suggestion.